### ISSUE TITLE
ggvis not rendering interactive plots with rstudio server

### ISSUE THREAD

**Issue by ewenharrison** [rstudio/ggvis#219]
Love the idea of ggvis as using shiny and ggplot a lot at the moment. But ggvis is not rendering interactive plots in RStudio-server .
- RStudio server version 0.98.953 running on ubuntu. 
- R 3.0.2
- Shiny-server 1.2.0.359 also running on same server.  
- ggvis 0.3.0.1
- shiny 0.10.0 
- Firefox 30.0 OR chrome 35.0.1916.153 m

This renders as expected:

```
mtcars %>%
  ggvis(~wt, ~mpg) %>%
  layer_points()
```

When trying to render the following, "viewer" window flashes but nothing appears. Message appears as expected:

```
Showing dynamic visualisation. Press Escape/Ctrl + C to stop

mtcars %>%
  ggvis(~wt, ~mpg) %>%
  layer_smooths(span = input_slider(0.5, 1, value = 1)) %>%
  layer_points(size := input_slider(100, 1000, value = 100))
```

ggvis plots on http://ggvis.rstudio.com/ etc. render as expected. 

Firefox or chrome both same. Appreciate any suggestions. Thank you!

**Comment by gourav-sg** [comment]
Hi,

I am working with nginx and included the following lines in `/etc/nginx/sites-enabled/default` for ubuntu 14.04, in order to resolve the issue:

```
        location / {
            proxy_pass http://127.0.0.1:8787;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
        }
```
The issue indeed had to do with enabling websocket via nginx

Regards,
Gourav

**Comment by slowsmile** [comment]
@ewenharrison, I encountered the same issue with you and I tried your solution yet it doesn't seem to work in my case. I also used apache2 as proxy to Rshiny server (port 8383)

when I run my R shiny app (in the directory of testapp under site_dir ) from firefox, the app interface can be shown then it stopped short with error message in error log:

Firefox can't establish a connection to the server at ws://mydomainname.edu/shinyport/testapp/__sockjs__/883/iy9jjfoj/websocket.

as I observed, the websocket didn't get to pass through with the original host name (www.mydomainname.edu) so I tried your solution, and added this in the apache2 configuration

<LocationMatch ^/testapp/_sockjs_/([0-9]+)/([a-z0-9]+)/websocket$>
                ProxyPassMatch ws://localhost:8383/testapp/_sockjs_/$1/$2/websocket
</LocationMatch>

The websocket changes on the two fields after "__sockjs__/" so I used regex to substitute them in ProxyPassMatch, however it still failed. I even confirmed the regex on a regular expression tester, but why it doesn't work in my case?

Any help is highly appreciated. It bothered me for a long while already.

**Comment by ewenharrison** [comment]
@slowsmile

You haven't provided enough information. Please provide the whole relevant
section from your apache2 configuration file. What regex expression are you
using? It doesn't look right. Are you using a <Location> directive? You
can't have "testapp" in ProxyPassMatch as that will change with each shiny
application. You could try this:

<Proxy *>
  Allow from localhost
</Proxy>

ProxyPass        / http://localhost:8383/
ProxyPassReverse / http://localhost:8383/
ProxyPassMatch   ^/shinyport/(.+)/websocket/$
ws://localhost:8383/shinyport/$1/websocket/

Untested. You may need to escape the / with \ in the regex, but probably
not. You will need to change ws to wss if you are using ssl/443.


On 23 June 2015 at 18:11, slowsmile <notifications@github.com> wrote:

> @ewenharrison <https://github.com/ewenharrison>, I encountered the same
> issue with you and I tried your solution yet it doesn't seem to work in my
> case. I also used apache2 as proxy to Rshiny server (port 8383)
>
> when I run my R shiny app (in the directory of testapp under site_dir )
> from firefox, the app interface can be shown then it stopped short with
> error message in error log:
>
> Firefox can't establish a connection to the server at ws://
> mydomainname.edu/shinyport/testapp/*sockjs*/883/iy9jjfoj/websocket.
>
> as I observed, the websocket didn't get to pass through with the original
> host name (www.mydomainname.edu) so I tried your solution, and added this
> in the apache2 configuration
>
>
> ProxyPassMatch ws://localhost:8383/testapp/*sockjs*/$1/$2/websocket
>
> The websocket changes on the two fields after "*sockjs*/" so I used regex
> to substitute them in ProxyPassMatch, however it still failed. I even
> confirmed the regex on a regular expression tester, but why it doesn't work
> in my case?
>
> Any help is highly appreciated. It bothered me for a long while already.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/ggvis/issues/219#issuecomment-114575136>.
>



-- 
Ewen M Harrison MB ChB PhD FRCS (Gen Surg)
Senior Lecturer in General Surgery | Consultant Transplant and
Hepatobiliary Surgeon
Director ChM General Surgery | GlobalSurg Research Collaborative
Surgical and Perioperative Health Research (SPHeRe)
Clinical Surgery
University of Edinburgh |  Royal Infirmary of Edinburgh | Edinburgh | EH16
4SA

Mobile: +44 (0) 797 442 0495
mail@ewenharrison.com
Web: University of Edinburgh
<http://www.ed.ac.uk/schools-departments/surgery/staff/surgical-profiles/ewen-harrison>
| DataSurg <http://datasurg.net> | ChM General Surgery
<http://www.chm.rcsed.ac.uk> | GlobalSurg <http://globalsurg.org>
Twitter: @ewenharrison
Skype: ewen.harrison

**Comment by slowsmile** [comment]
@ewenharrison 
Thanks for your response and followed your suggestions, but the error message remains. Here is the section in my apache config file



        <Directory />
        #Options FollowSymLinks
        Options Indexes FollowSymLinks Includes ExecCGI
        AllowOverride All
        #    Order deny,allow
        #    Allow from all
             Require all granted
        </Directory>


        ProxyRequests On
        ProxyVia On

        <Proxy *>
            Require all granted
        </Proxy>

        ProxyPass /shinyport/ http://localhost:8383/
        ProxyPassReverse /shinyport/ http://localhost:8383/
        RedirectMatch permanent ^/shinyport$ /shinyport/
        ProxyPassMatch ^/shinyport/(.+)/websocket$ wss://localhost:8383/$1/websocket


        # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
        # error, crit, alert, emerg.
        # It is also possible to configure the loglevel for particular
        # modules, e.g.
        #LogLevel info ssl:warn

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined


Basically, I forward to port 8383 from default 80 if the user opens www.mydomainname.edu/shinyport  and there are several R-shiney apps in /shinyport including /BrainSpanapp

I was able to open the app panel but the websocket always failed to connect. Here is the message again


Firefox can't establish a connection to the server at ws://mydomainname.edu/shinyport/BrainSpanapp/__sockjs__/912/8_iofafc/websocket. 

Any additional suggestion is highly appreciated!

**Comment by ewenharrison** [comment]
Hi,


It's difficult without being able to test it I'm afraid. The ProxyPassMatch
statement doesn't look like it will owrk as it misses out /shinyport/ from
the wss address. My last suggestion:
Replace:

ProxyPass /shinyport/ http://localhost:8383/
ProxyPassReverse /shinyport/ http://localhost:8383/
RedirectMatch permanent ^/shinyport$ /shinyport/
ProxyPassMatch ^/shinyport/(.+)/websocket$ wss://localhost:8383/$1/websocket

with:

    <Location /shinyport>
        ProxyPass        http://localhost:8383/shinyport
        ProxyPassReverse    http://localhost:3838/shinyport
    </Location>

    <LocationMatch ^/shinyport/(.*)/websocket$>
        ProxyPassMatch wss://localhost:8383/shinyport/$1/websocket
    </LocationMatch>

Make sure in /etc/shiny-server/shiny-server.conf you have:
location /shinyport {
}

Ewen

On 24 June 2015 at 17:37, slowsmile <notifications@github.com> wrote:

> @ewenharrison <https://github.com/ewenharrison>
> Thanks for your response and followed your suggestions, but the error
> message remains. Here is the section in my apache config file
>
>     <Directory />
>     #Options FollowSymLinks
>     Options Indexes FollowSymLinks Includes ExecCGI
>     AllowOverride All
>     #    Order deny,allow
>     #    Allow from all
>          Require all granted
>     </Directory>
>
>
>     ProxyRequests On
>     ProxyVia On
>
>     <Proxy *>
>         Require all granted
>     </Proxy>
>
>     ProxyPass /shinyport/ http://localhost:8383/
>     ProxyPassReverse /shinyport/ http://localhost:8383/
>     RedirectMatch permanent ^/shinyport$ /shinyport/
>     ProxyPassMatch ^/shinyport/(.+)/websocket$ wss://localhost:8383/$1/websocket
>
>
>     # Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
>     # error, crit, alert, emerg.
>     # It is also possible to configure the loglevel for particular
>     # modules, e.g.
>     #LogLevel info ssl:warn
>
>     ErrorLog ${APACHE_LOG_DIR}/error.log
>     CustomLog ${APACHE_LOG_DIR}/access.log combined
>
> Basically, I forward to port 8383 from default 80 if the user opens
> www.mydomainname.edu/shinyport and there are several R-shiney apps in
> /shinyport including /BrainSpanapp
>
> I was able to open the app panel but the websocket always failed to
> connect. Here is the message again
>
> Firefox can't establish a connection to the server at ws://
> mydomainname.edu/shinyport/BrainSpanapp/*sockjs*/912/8_iofafc/websocket.
>
> Any additional suggestion is highly appreciated!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rstudio/ggvis/issues/219#issuecomment-114936229>.
>



-- 
Ewen M Harrison MB ChB PhD FRCS (Gen Surg)
Senior Lecturer in General Surgery | Consultant Transplant and
Hepatobiliary Surgeon
Director ChM General Surgery | GlobalSurg Research Collaborative
Surgical and Perioperative Health Research (SPHeRe)
Clinical Surgery
University of Edinburgh |  Royal Infirmary of Edinburgh | Edinburgh | EH16
4SA

Mobile: +44 (0) 797 442 0495
mail@ewenharrison.com
Web: University of Edinburgh
<http://www.ed.ac.uk/schools-departments/surgery/staff/surgical-profiles/ewen-harrison>
| DataSurg <http://datasurg.net> | ChM General Surgery
<http://www.chm.rcsed.ac.uk> | GlobalSurg <http://globalsurg.org>
Twitter: @ewenharrison
Skype: ewen.harrison
